### PR TITLE
[HLSL] Add -HV option translation to clang-dxc.exe

### LINKED ISSUE #83479: [HLSL] Implement -HV option for DXC
Currently, the -HV option that is typically taken by the DXC compiler isn't properly translated to an appropriate set of options for clang in -cc1 mode. This issue requires that an appropriate option translation is implemented.

AC:
Running:
`"clang-dxc.exe" -enable-16bit-types -HV 2016 -T lib_6_1 <test hlsl file> -###`
Should result in a command line to clang-cc1 that includes a set of options that has equivalent semantic meaning to adding -HV to the DXC compiler.

### PR DESCRIPTION
**Pull request by bob80905**
Previously, clang-dxc.exe would not recognize -HV as a valid argument to DXC, and would be unable to translate the argument to a legal clang argument. This PR implements a translation of the HV option and its value to the appropriate clang flag and the appropriate value. It adds a test by using the -### option to spit out the translated options, and checks to see that the correct option was generated.
Fixes #83479

### CODE REVIEW

File: clang/lib/Driver/ToolChains/HLSL.cpp [hunk 1]
```diff
@@ -226,6 +226,28 @@ HLSLToolChain::TranslateArgs(const DerivedArgList &Args, StringRef BoundArch,
       A->claim();
       continue;
     }
+    if (A->getOption().getID() == options::OPT_dxc_hlsl_version) {
+      // Translate -HV into -std for llvm
+      // depending on the value given, assign std to:
+      // c++14,c++17,c++20,c++latest,c11,c17
```
Comment (llvm-beanz) [inline clang/lib/Driver/ToolChains/HLSL.cpp:232]: I think your comment here is wrong.

File: clang/lib/Driver/ToolChains/HLSL.cpp [hunk 2]
```diff
@@ -226,6 +226,28 @@ HLSLToolChain::TranslateArgs(const DerivedArgList &Args, StringRef BoundArch,
...
+      } else if (strcmp(value, "2017") == 0) {
+        DAL->AddSeparateArg(nullptr, Opts.getOption(options::OPT_std_EQ),
+                            "hlsl2017");
+      } else if (strcmp(value, "2018") == 0) {
+        DAL->AddSeparateArg(nullptr, Opts.getOption(options::OPT_std_EQ),
+                            "hlsl2018");
+      } else if (strcmp(value, "2021") == 0) {
+        DAL->AddSeparateArg(nullptr, Opts.getOption(options::OPT_std_EQ),
+                            "hlsl2021");
+      }
```
Comment (llvm-beanz) [inline clang/lib/Driver/ToolChains/HLSL.cpp:246]: What about 202x? What about unrecognized values?

**Review by llvm-beanz — COMMENTED** [review]
I worry this pattern will also lend to us having multiple places that parse and translate HLSL versions. Can you look at other places in Clang where we handle these numbers and see if there is a good way to unify the string handling? It would be ideal for us to only have one place we need to modify when we add new language versions.

File: clang/include/clang/Driver/Options.td [hunk 3]
```diff
@@ -8545,6 +8545,10 @@ def dxc_entrypoint : Option<["--", "/", "-"], "E", KIND_JOINED_OR_SEPARATE>,
                      Group<dxc_Group>,
                      Visibility<[DXCOption]>,
                      HelpText<"Entry point name">;
+def dxc_hlsl_version : Option<["--", "/", "-"], "HV", KIND_JOINED_OR_SEPARATE>,
+                     Group<dxc_Group>,
+                     Visibility<[DXCOption]>,
+                     HelpText<"HLSL Version">;
```
Comment (bogner) [inline clang/include/clang/Driver/Options.td:8551]: This doesn't look right - I don't think `--HV` is valid.

File: clang/test/Options/HV.hlsl [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// RUN: %clang_dxc -T lib_6_4 -HV 2016 %s 2>&1 -###   | FileCheck -check-prefix=2016 %s
+// RUN: %clang_dxc -T lib_6_4 -HV 2017 %s 2>&1 -###   | FileCheck -check-prefix=2017 %s
+// RUN: %clang_dxc -T lib_6_4 -HV 2018 %s 2>&1 -###   | FileCheck -check-prefix=2018 %s
+// RUN: %clang_dxc -T lib_6_4 -HV 2021 %s 2>&1 -###   | FileCheck -check-prefix=2021 %s
```
Comment (bogner) [inline clang/test/Options/HV.hlsl:4]: Please also include at least one test of the `/HV` spelling

File: clang/lib/Driver/ToolChains/HLSL.cpp (hunk 2 above)
Comment (bob80905) [reply clang/lib/Driver/ToolChains/HLSL.cpp:265]: Added a case for 202x and for unrecognized values. 
I notice 2015 used to be an accepted value, but I'm not sure how we want to treat that option, since DXC seems to treat it as a special option and has extra conditions for accepting it.

File: clang/include/clang/Driver/Options.td [hunk 5]
```diff
@@ -8545,6 +8545,11 @@ def dxc_entrypoint : Option<["--", "/", "-"], "E", KIND_JOINED_OR_SEPARATE>,
                      Group<dxc_Group>,
                      Visibility<[DXCOption]>,
                      HelpText<"Entry point name">;
+def dxc_hlsl_version : Option<["/", "-"], "HV", KIND_JOINED_OR_SEPARATE>,
+                     Group<dxc_Group>,
+                     Visibility<[DXCOption]>,
+                     HelpText<"HLSL Version">,
+                     Values<"2016, 2017, 2018, 2021, 202x">;
```
Comment (python3kgae) [inline clang/include/clang/Driver/Options.td:8552]: Does the Values already make sure error be reported when give input not in the Values?

Comment (bob80905) [reply clang/include/clang/Driver/Options.td:8552]: No, I added that logic myself.  I've manually checked all the valid values and emit an error if there are no matches in  HLSL.cpp:263.
Regardless, I wouldn't need that logic, because I need to perform a unique transformation exhaustively in every valid case, and if there are no matching valid cases, then I emit an error with `else`.

File: clang/lib/Driver/ToolChains/HLSL.cpp [hunk 6]
```diff
@@ -226,6 +226,47 @@ HLSLToolChain::TranslateArgs(const DerivedArgList &Args, StringRef BoundArch,
...
     }
+    if (A->getOption().getID() == options::OPT_dxc_hlsl_version) {
+      // Translate -HV into -std for llvm
+      // depending on the value given
+      llvm::StringRef value(A->getValue());
+      llvm::StringRef acceptedValues(
+          Opts.getOptionValues(options::OPT_dxc_hlsl_version));
+      llvm::SmallVector<StringRef, 8> validValues;
+      acceptedValues.split(validValues, ", ");
+      if (value == validValues[0]) {
```
Comment (python3kgae) [inline clang/lib/Driver/ToolChains/HLSL.cpp:237]: Why not just check value == "2016" here so we don't need to getOptionValues?
And the code will be easier to understand.

Comment (bob80905) [reply clang/lib/Driver/ToolChains/HLSL.cpp:237]: The idea to use getOptionValues was an idea inspired by @llvm-beanz 's feedback, to prevent littering direct string comparisons into the code and unify string handling. Originally, I did have the direct "2016" comparison. I'm not sure if the LLVM community would want access to these option values, but it seems like the most convenient way to get around direct string comparisons. I do remember string comparisons were a pain point in the DXC codebase and want to try and avoid that.

Comment (python3kgae) [reply clang/lib/Driver/ToolChains/HLSL.cpp:237]: We can do something like https://github.com/llvm/llvm-project/blob/main/clang/lib/Frontend/CompilerInvocation.cpp#L3689 where the OPT_std_EQ is checked 
 with the LangStandard::getLangKind at https://github.com/llvm/llvm-project/blob/main/clang/lib/Basic/LangStandards.cpp#L64

Maybe add LangStandard::getHLSLLangKind ?

Comment (bob80905) [reply clang/lib/Driver/ToolChains/HLSL.cpp]: I've added your recommended function, hopefully it is what you envisioned?

File: clang/lib/Basic/LangStandards.cpp [hunk 7]
```diff
@@ -69,6 +69,21 @@ LangStandard::Kind LangStandard::getLangKind(StringRef Name) {
       .Default(lang_unspecified);
 }
 
+LangStandard::Kind LangStandard::getHLSLLangKind(StringRef Name) {
+  if (Name == "2016")
```
Comment (python3kgae) [inline clang/lib/Basic/LangStandards.cpp:73]: Maybe a StringSwitch here.

**Review by python3kgae — APPROVED** [review]

File: clang/include/clang/Driver/Options.td (hunk 5 above)
Comment (llvm-beanz) [reply clang/include/clang/Driver/Options.td:8552]: I think the problem is that you're using `Values` wrong. That should be taking a list of possible values not a string.

Look at how the `target_profile` flag is implemented for an example.

Comment (llvm-beanz) [reply clang/include/clang/Driver/Options.td:8552]: Oh! Also look at some of the other options. You can use a combination of `Values`, `NormalizedValues` and `MarshallingInfoEnum`, to have the parser generated to parse directly to the `HLSLLangStd` enumeration.

Comment (bob80905) [reply clang/include/clang/Driver/Options.td:8552]: I've decided to use NormalizedValues instead of values, but I no longer need access to valid values thanks to Xiang's suggestion. So I think I should just leave the values there for documentation's sake, but the values won't actually be read by option translation.

File: clang/lib/Basic/LangStandards.cpp [hunk 8]
```diff
@@ -69,6 +69,16 @@ LangStandard::Kind LangStandard::getLangKind(StringRef Name) {
...
 
+LangStandard::Kind LangStandard::getHLSLLangKind(StringRef Name) {
+  return llvm::StringSwitch<LangStandard::Kind>(Name)
+      .Case("2016", LangStandard::lang_hlsl2016)
+      .Case("2017", LangStandard::lang_hlsl2017)
+      .Case("2018", LangStandard::lang_hlsl2018)
+      .Case("2021", LangStandard::lang_hlsl2021)
+      .Case("202x", LangStandard::lang_hlsl202x)
+      .Default(LangStandard::lang_unspecified);
+}
```
Comment (llvm-beanz) [inline clang/lib/Basic/LangStandards.cpp:80]: This function isn't needed if you use the Options Enum marshalling that I suggested.

File: clang/include/clang/Driver/Options.td [hunk 9]
```diff
@@ -8549,7 +8549,10 @@ def dxc_hlsl_version : Option<["/", "-"], "HV", KIND_JOINED_OR_SEPARATE>,
                      Group<dxc_Group>,
                      Visibility<[DXCOption]>,
                      HelpText<"HLSL Version">,
-                     NormalizedValues<["2016", "2017", "2018", "2021", "202x"]>;
+                     Values<"2016, 2017, 2018, 2021, 202x">,
+                     NormalizedValues<["lang_hlsl2016", "lang_hlsl2017", "lang_hlsl2018", "lang_hlsl2021", "lang_hlsl202x"]>,
+                     NormalizedValuesScope<"LangStandard">,
+                     MarshallingInfoEnum<LangOpts<"HLSLVersion">, "lang_hlsl2021">;
```
Comment (llvm-beanz) [inline clang/include/clang/Driver/Options.td:8555]: Do we need a default value for this?
The default language version is set elsewhere to HLSL 2021, having a default here will make it so that we need to set the default in two places instead of one when we change it.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 10]
```diff
@@ -753,6 +753,9 @@ def err_drv_hlsl_bad_shader_required_in_target : Error<
 
 def err_drv_hlsl_bad_shader_unsupported : Error<
   "%select{shader model|Vulkan environment|shader stage}0 '%1' in target '%2' is invalid for HLSL code generation">;
+def err_drv_hlsl_dxc_bad_argument_value : Error<
+  "'%0' dxc argument has invalid value '%1'">;
+
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:758]: This doesn't seem to be used anymore.

File: clang/include/clang/Basic/LangStandard.h [hunk 11]
```diff
@@ -139,6 +139,7 @@ struct LangStandard {
   bool isOpenCL() const { return Flags & OpenCL; }
 
   static Kind getLangKind(StringRef Name);
+  static Kind getHLSLLangKind(StringRef Name);
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/LangStandard.h:142]: This either.

File: clang/lib/Driver/ToolChains/HLSL.cpp [hunk 12]
```diff
@@ -226,6 +226,7 @@ HLSLToolChain::TranslateArgs(const DerivedArgList &Args, StringRef BoundArch,
       A->claim();
       continue;
     }
+
```
Comment (llvm-beanz) [inline clang/lib/Driver/ToolChains/HLSL.cpp:229]: Unnecessary whitespace change.

File: clang/test/Options/HV.hlsl [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+// RUN: %clang_dxc -T lib_6_4 /HV 2021 %s 2>&1 -###   | FileCheck -check-prefix=2021 %s
+// RUN: %clang_dxc -T lib_6_4 /HV 202x %s 2>&1 -###   | FileCheck -check-prefix=202x %s
+// RUN: not %clang_dxc -T lib_6_4 /HV gibberish -### %s 2>&1 | FileCheck -check-prefix=CHECK-ERR %s
+
+// 2016: "-std=hlsl2016"
+// 2017: "-std=hlsl2017"
+// 2018: "-std=hlsl2018"
+// 2021: "-std=hlsl2021"
+// 202x: "-std=hlsl202x"
+// CHECK-ERR: error: 'HV' dxc argument has invalid value 'gibberish'
```
Comment (llvm-beanz) [inline clang/test/Options/HV.hlsl:13]: We should be able to test that the `-std` flag is set for each possible value (as you do), and that if you don't pass `/HV` we don't pass through a `-std` flag.

File: clang/include/clang/Driver/Options.td [hunk 14]
```diff
@@ -8545,6 +8545,11 @@ def dxc_entrypoint : Option<["--", "/", "-"], "E", KIND_JOINED_OR_SEPARATE>,
                      Group<dxc_Group>,
                      Visibility<[DXCOption]>,
                      HelpText<"Entry point name">;
+def dxc_hlsl_version : Option<["/", "-"], "HV", KIND_JOINED_OR_SEPARATE>,
+                     Group<dxc_Group>,
+                     Visibility<[DXCOption]>,
+                     HelpText<"HLSL Version">,
+                     NormalizedValues<["2016", "2017", "2018", "2021", "202x"]>;
```
Comment (llvm-beanz) [inline clang/include/clang/Driver/Options.td:8552]: In the current state of this patch I don't think this does anything. I don't think the `Values` or `NormalizedValues` stuff really do anything (other than maybe populating the help spew) if you're not using the marshalling infrastructure (https://clang.llvm.org/docs/InternalsManual.html#option-marshalling-infrastructure).

We don't currently have a `DriverOptions` construct, so driver-only options can't use the marshalling stuff, which is unfortunate.

I think for the purposes of this change we should probably go with the direction in this PR and revisit marshalling later.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 15]
```diff
@@ -753,6 +753,9 @@ def err_drv_hlsl_bad_shader_required_in_target : Error<
 
 def err_drv_hlsl_bad_shader_unsupported : Error<
   "%select{shader model|Vulkan environment|shader stage}0 '%1' in target '%2' is invalid for HLSL code generation">;
+def err_drv_hlsl_dxc_bad_argument_value : Error<
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:756]: Can we use `err_drv_invalid_value` instead? I don't think there is any reason the error needs to say `dxc` in the text.

**Review by llvm-beanz — COMMENTED** [review]
Minor suggestion about the diagnostic.

@MaskRay do you have any thoughts about a follow-up change to add a `DriverOptions` structure so that we could use argument marshalling for driver options?

File: clang/include/clang/Driver/Options.td (hunk 14 above)
Comment (bob80905) [reply clang/include/clang/Driver/Options.td:8552]: Populating the help spew seems helpful! I can remove the marshalling part, but it seems helpful to leave the rest as is.

File: clang/include/clang/Basic/DiagnosticDriverKinds.td [hunk 16]
```diff
@@ -753,6 +753,7 @@ def err_drv_hlsl_bad_shader_required_in_target : Error<
 
 def err_drv_hlsl_bad_shader_unsupported : Error<
   "%select{shader model|Vulkan environment|shader stage}0 '%1' in target '%2' is invalid for HLSL code generation">;
+
```
Comment (llvm-beanz) [inline clang/include/clang/Basic/DiagnosticDriverKinds.td:756]: nit: leftover whitespace change.
```suggestion
```

File: clang/include/clang/Driver/Options.td [hunk 17]
```diff
@@ -8545,6 +8545,13 @@ def dxc_entrypoint : Option<["--", "/", "-"], "E", KIND_JOINED_OR_SEPARATE>,
                      Group<dxc_Group>,
                      Visibility<[DXCOption]>,
                      HelpText<"Entry point name">;
+def dxc_hlsl_version : Option<["/", "-"], "HV", KIND_JOINED_OR_SEPARATE>,
+                     Group<dxc_Group>,
+                     Visibility<[DXCOption]>,
+                     HelpText<"HLSL Version">,
+                     Values<"2016, 2017, 2018, 2021, 202x">,
+                     NormalizedValuesScope<"LangStandard">,
+                     NormalizedValues<["hlsl2016", "hlsl2017", "hlsl2018", "hlsl2021", "hlsl202x"]>;
```
Comment (llvm-beanz) [inline clang/include/clang/Driver/Options.td:8554]: nit: Do we need these? The `-help` spew should be driven by the `Values` entry right?
```suggestion
```

**Review by llvm-beanz — APPROVED** [review]
Two small nits with suggestions. Otherwise LGTM.
